docs/rfcs/001-project-specific-configuration.md

```diff
@@ -0,0 +1,126 @@
+
+## Motivation
+
+People who use Atom often work on multiple projects. Sometimes these projects have very different requirements for code formatting, tools to use, or other features that will necessitate different configuration of the editing environment. Currently, if one needs different settings for different projects, one has to use a community package, use [editorconfig](http://editorconfig.org/) which has a limited set of configuration values, or resort to manually changing configuration each time one switches to a different project.
+
+## Explanation
+
+There are two configuration files: the global `~/.atom/config.cson` and the project-specific `atom-config.cson`. (Hereafter these two files will be referred to as the "global settings" and "project settings" respectively.) The project settings file is stored in the root directory of the project root.
+
+When reading from the configuration using `atom.config.get()`, the contents of the two configuration files is merged using [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) or some similar mechanism, with the values in the project settings taking precedence over the global settings. For example:
+
+```coffee
+# Global settings contents
+editor:
+  fontSize: 14
+```
+
+```coffee
+# Project settings contents
+editor:
+  fontSize: 16
+```
+
+```javascript
+let fontSize = atom.config.get('editor.fontSize')
+console.log(`fontSize = ${fontSize}`)
+// Outputs "fontSize = 16"
+```
+
+When setting or unsetting a configuration value, `atom.config.set()` and `atom.config.unset()` will continue to work the way it always has unless the `project` key is set to `true` in the options parameter. For example:
+
+```javascript
+// Sets the global editor.fontSize value to 18
+atom.config.set('editor.fontSize', 18)
+
+// Sets the project editor.fontSize value to 22
+atom.config.set('editor.fontSize', 22, {project: true})
+
+// Unsets the project core.fileEncoding value, leaving the global one unchanged
+atom.config.unset('core.fileEncoding', {project: true})
+```
+
+Setting a language-specific configuration can also be done at either the global or project setting level:
+
+```javascript
+// Sets the editor.tabLength setting to 4 for GitHub-flavored Markdown files globally
+atom.config.set('editor.tabLength', 4, {scopeSelector: '.source.gfm'})
+
+// Sets the editor.tabLength setting to 2 for GitHub-flavored Markdown files in this project
+atom.config.set('editor.tabLength', 2, {project: true, scopeSelector: '.source.gfm'})
+```
+
+**Note:** [Behavior is undefined](https://blogs.msdn.microsoft.com/oldnewthing/20140627-00/?p=633) when there are multiple project roots that have `atom-config.cson` files whose individual values are different. For example, given a project that has two project roots "foo" and "bar" with project settings like the following:
```

Comment:
Should this fall back to the `global` value instead?

Reply:
Just to be clear, leaving it `undefined` is valid at this point as far as I'm concerned.

Reply:
Leaving this alone for now.